src/libraries/Microsoft.Extensions.Options/gen/Generator.cs

```diff
@@ -0,0 +1,47 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.Extensions.Options.Generators
+{
+    [Generator]
+    public class Generator : IIncrementalGenerator
+    {
+        private static readonly HashSet<string> _attributeNames = new()
+        {
+            SymbolLoader.OptionsValidatorAttribute,
+        };
+
+        public void Initialize(IncrementalGeneratorInitializationContext context)
+        {
+            GeneratorUtilities.Initialize(context, _attributeNames, HandleAnnotatedTypes);
+        }
+
+        private static void HandleAnnotatedTypes(Compilation compilation, IEnumerable<SyntaxNode> nodes, SourceProductionContext context)
+        {
+            if (!SymbolLoader.TryLoad(compilation, out var symbolHolder))
+            {
+                // Not eligible compilation
+                return;
+            }
+
+            var parser = new Parser(compilation, context.ReportDiagnostic, symbolHolder!, context.CancellationToken);
```

Comment:
Following up on https://github.com/dotnet/runtime/pull/87587/files#r1231203554, the `Parser` class should not take a dependency on `SourceProductionContext`. The pattern typically followed is to cache `Diagnostics` in an [incremental value](https://github.com/dotnet/runtime/blob/1ff48ad53e01d1db3b5ad6b247347dc5609cb93a/src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/LibraryImportGenerator.cs#L82-L101) so that they can be reported if and when the incremental value is passed to a `SourceProductionContext`.